services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+    if (client_state_ != kStopped) {
+      StopProvider();
+    }
+    high_accuracy_requested_ = true;
+  }
+
+  if (client_state_ != kStopped) {
+    return;
+  }
+  client_state_ = kInitializing;
+
+  dbus::ObjectProxy* proxy =
+      bus_->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  proxy->CallMethod(
+      &call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+      base::BindOnce(&GeoClueLocationProvider::OnGetClientCompleted,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueLocationProvider::StopProvider() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (client_state_ == kStopped) {
+    return;
+  }
+
+  client_state_ = kStopped;
+
+  // Invalidate weak pointers, so we don't continue any async operations.
+  weak_ptr_factory_.InvalidateWeakPtrs();
+
+  // Stop can be called before the gclue_client_ has resolved.
+  if (!gclue_client_) {
+    return;
+  }
+
+  dbus::MethodCall stop(kClientInterfaceName, "Stop");
+  gclue_client_->CallMethod(&stop, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+                            base::DoNothing());
+
+  // Reset pointers to dbus objects.
+  gclue_client_.reset();
+}
+
+const mojom::Geoposition& GeoClueLocationProvider::GetPosition() {
+  return last_position_;
+}
+
+void GeoClueLocationProvider::OnPermissionGranted() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  permission_granted_ = true;
+  StartClient();
+}
+
+void GeoClueLocationProvider::SetPosition(const mojom::Geoposition& position) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  last_position_ = position;
+
+  if (client_state_ != kStarted) {
+    return;
+  }
+
+  if (last_position_.error_code == mojom::Geoposition_ErrorCode::NONE &&
+      !device::ValidateGeoposition(last_position_)) {
+    return;
+  }
+  position_update_callback_.Run(this, last_position_);
+}
+
+void GeoClueLocationProvider::OnGetClientCompleted(dbus::Response* response) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!response) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  dbus::MessageReader reader(response);
+  dbus::ObjectPath path;
+  if (!reader.PopObjectPath(&path)) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  gclue_client_ = bus_->GetObjectProxy(kServiceName, path);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  auto* raw_ptr = new GeoClueProperties(gclue_client_.get());
+  std::unique_ptr<dbus::PropertySet> properties(raw_ptr);
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr(), std::move(properties)));
+
+  raw_ptr->requested_accuracy_level.Set(
```

Comment:
Reusing the underlying pointer after make_unique seems bad, too?

Reply:
yea, the usage here is backwards and you should get the raw ptr from the unique_ptr, but also `raw_ptr` is not a great variable name. raw_ptr to what?